src/mds/MDSRank.cc

```diff
@@ -3373,7 +3351,7 @@ bool MDSRank::evict_client(int64_t session_id,
 
     Context *on_blacklist_done = new FunctionContext([this, fn](int r) {
       objecter->wait_for_latest_osdmap(
-       new C_OnFinisher(
+      std::ref(*(new C_OnFinisher(
```

Comment:
I'm somewhat confused looking at this as to the expected lifetime of the anonymous C_OnFinisher. Asyncronous magic, or slightly dangerous looking..?

Reply:
The C_OnFinisher deletes itself when called so it's safe, but /ugly/.

The motivation being to limit the degree to which I had to rewrite everything in the world with every change. Though I became more conservative and started providing interfaces to Objecter that used Context* and wrapped it in a lambda. Initially I was hoping to use this as an excuse to cut down the use of Context* all over Ceph, but that ended up turning into a huge amount of scope creep.